plugins/devlink/fu-devlink-device.c

```diff
@@ -451,24 +418,13 @@ fu_devlink_device_info_cb(const struct nlmsghdr *nlh, gpointer data)
 		fu_device_set_serial(device, serial_number);
 	}
 
-	version_table = fu_devlink_device_get_version_table(nlh);
-
-	/* remove components that are not in the version table */
-	for (guint i = 0; i < children->len; i++) {
-		FuDevice *component = g_ptr_array_index(children, children->len - i - 1);
-		const gchar *name = fu_device_get_name(component);
-		FuDevlinkVersionInfo *version_info = g_hash_table_lookup(version_table, name);
-
-		if (version_info == NULL ||
-		    (version_info->stored == NULL && version_info->running == NULL)) {
-			g_debug("removed component %s", name);
-			fu_device_remove_child(device, component);
-		}
-	}
-
+	attrs_map = fu_devlink_device_populate_attrs_map(nlh);
 	helper.device = device;
-	helper.version_table = version_table;
-	g_hash_table_foreach(version_table, fu_devlink_device_update_component_cb, &helper);
+	helper.psid = g_hash_table_lookup(attrs_map, "fw.psid");
+	helper.version_bootloader = g_hash_table_lookup(attrs_map, "fw.bootloader");
```

Comment:
bootloader (if any) is a separate component from kernel perspective.

Reply:
not for current kernel driver. But in theory, why not? The point is devlink assumes version-component to be in 1:1 relationship. The strings exposed in devlink info and accepted by devlink flash are the same.

Reply:
> The fwupd device model has the bootloader version (assumed to be r/o) as a property on the main FuDevice. I think if there is a device that supports updating the bootloader we can certainly add a quirk for it then, but until then it seems a very dangerous thing to do _in the field_ so to speak.

In that case, the bootloader version might get special treating and just be used for the root devlink device set (similar to how it is done with serial number) and ignored for component creation until there is a device to support it with a special quirk. How does that sound?